src/io_callbacks.zig

```diff
@@ -0,0 +1,902 @@
+                    linux.io_uring_prep_fsync(sqe, op.fd, op.flags);
+                },
+                .openat => |op| {
+                    linux.io_uring_prep_openat(sqe, op.fd, op.path, op.flags, op.mode);
+                },
+                .read => |op| {
+                    linux.io_uring_prep_read(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+                .recv => |op| {
+                    linux.io_uring_prep_recv(sqe, op.socket, op.buffer, op.flags);
+                },
+                .send => |op| {
+                    linux.io_uring_prep_send(sqe, op.socket, op.buffer, op.flags);
+                },
+                .write => |op| {
+                    linux.io_uring_prep_write(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+            }
+            sqe.user_data = @ptrToInt(completion);
+        }
+
+        fn complete(completion: *Completion) void {
+            switch (completion.operation) {
+                .accept => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
+                        os.EINTR => {
+                            completion.io.enqueue(completion);
+                            return;
+                        },
+                        os.EAGAIN => error.WouldBlock,
+                        os.EBADF => error.FileDescriptorInvalid,
+                        os.ECONNABORTED => error.ConnectionAborted,
+                        os.EFAULT => unreachable,
+                        os.EINVAL => error.SocketNotListening,
+                        os.EMFILE => error.ProcessFdQuotaExceeded,
+                        os.ENFILE => error.SystemFdQuotaExceeded,
+                        os.ENOBUFS => error.SystemResources,
+                        os.ENOMEM => error.SystemResources,
+                        os.ENOTSOCK => error.FileDescriptorNotASocket,
+                        os.EOPNOTSUPP => error.OperationNotSupported,
+                        os.EPERM => error.PermissionDenied,
+                        os.EPROTO => error.ProtocolFailure,
+                        else => |errno| os.unexpectedErrno(@intCast(usize, errno)),
+                    } else @intCast(os.socket_t, completion.result);
+                    completion.callback(completion.context, completion, &result);
+                },
+                .close => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
+                        os.EINTR => {}, // A success, see https://github.com/ziglang/zig/issues/2425.
+                        os.EBADF => error.FileDescriptorInvalid,
+                        os.EDQUOT => error.DiskQuota,
+                        os.EIO => error.InputOutput,
+                        os.ENOSPC => error.NoSpaceLeft,
+                        else => |errno| os.unexpectedErrno(@intCast(usize, errno)),
+                    } else assert(completion.result == 0);
```

Comment:
What is the value of `const result =` on line 219 if we only have an `assert()` on `completion.result` here on line 226 without providing anything explicitly (i.e. as we would normally do on line 215)?

Reply:
I realize now that this probably isn't as clear as it was in my head. The return type of `assert()` is `void` so result (which is of type `CloseError!void`) would be assigned the value `void` in this case to indicate that there was no error.

Reply:
It would be more explicit to do
```zig
blk: {
    assert(completion.result == 0);
    break :blk {};
};
```

Reply:
> I realize now that this probably isn't as clear as it was in my head. The return type of assert() is void so result (which is of type CloseError!void) would be assigned the value void in this case to indicate that there was no error.

No, you're right. I had the 0/void type mixed up. Let's leave it as is, or as you prefer.